### ISSUE TITLE
add mac address randomization

### ISSUE THREAD

**Issue by Bogdan107** [v1s1t0r1sh3r3/airgeddon#448]
## Example use case:
1) Interface wlan0 controlled by NetworkManager by default and used as test client with test AP.
2) Interface wlan1 configuread as unmanaged in NetworkManager and used as attacking device.
```
cat /etc/NetworkManager/NetworkManager.conf
[keyfile]
unmanaged-devices=interface-name:wlan1
```
3) start airgeddon;
4) select "wps attack";
5) start attack;
6) Attack always start with default/permanent mac address of interface.

### 1) "MAC Randomization feature"
Realized by external [macchanger](https://github.com/alobbs/macchanger) application. For example:
```
ifconfig wlan1 down
macchanger -r wlan1
ifconfig wlan1 up
```

### 2) Random MAC-address after each "Interface selection" stage.
MAC-address changing must automatically do after each choise of "Select an interface".

Feature may be controlled by configuration option like AIRGEDDON_AUTO_MAC_CHANGE in airgeddonrc file.

### 3) Random MAC-address before each attack.
If attacking application has no mac-spoof feature, than MAC-address must be spoofed before each attack.
This is actual for "WPS attack/bully", which always say which mac-address used for attack in line, like this: "Using 'XX:XX:XX:X:XX:XX' for the source MAC address". And this mac-address always the same as original (permanent, manufactured) of interface adapter, if this address don't changed manually.

Feature may be controlled by configuration option like AIRGEDDON_AUTO_MAC_CHANGE in airgeddonrc file.

## What are the benefits of adding this?

Forced MAC-spoofing feature:
- allows to hide attack from security systems, which do not alerts by single WPS-pairing attempt with unique MAC-address, as it may be legal;
- helps to tune security system triggers.

**Comment by OscarAkaElvis** [comment]
For now, mac spoofing is only available for all Evil Twin attacks. airgeddon ask to the user to perform it. And we are using a different method than macchanger to perform it. Doing it using macchanger is another unnecessary dependency. In airgeddon it is done at function called `set_spoofed_mac` in a Linux native way without needing any external tool.

Thanks for the suggestion but it sounds to me like an option to be created using a plugin. Do you know how to do a plugin? is pretty simple, you can do it yourself. Check the plugins system: https://github.com/v1s1t0r1sh3r3/airgeddon/wiki/Plugins%20System

I'll close this for now. Who knows, maybe in the future we can do something like that.

**Comment by Bogdan107** [comment]
Plugin [proposed](https://github.com/v1s1t0r1sh3r3/airgeddon/compare/master...Bogdan107:patch-1#diff-7bf0aec8ccf62879f62e946238160952d064fd97f2863b24941aab160fbec13b)

**Comment by xpz3** [comment]
@OscarAkaElvis take a look. Nice feature I think. Maybe add it to plugin hall of fame

> Plugin [proposed](https://github.com/v1s1t0r1sh3r3/airgeddon/compare/master...Bogdan107:patch-1#diff-7bf0aec8ccf62879f62e946238160952d064fd97f2863b24941aab160fbec13b)

Good one @Bogdan107

**Comment by OscarAkaElvis** [comment]
Yeah, you got how to make a plugin very fast! awesome!

Checking your plugin, just a suggestion... on evil twin attacks, the possibility to spoof mac is already existing asking to the user... now your plugin is forcing it, but the user still will be asked about it. Maybe you could override function `et_prerequisites`, copy and paste all the content but removing the part:

```
ask_yesno 419 "no"
	if [ "${yesno}" = "y" ]; then
		mac_spoofing_desired=1
	fi
```

That will make the question to the user to be avoided. Just a suggestion. Anyway, good job!

If you do more plugins, you can be linked in our wiki at plugins hall of fame section.

I can suggest you to join our Discord channel for further discussion. This is the link: https://discord.gg/sQ9dgt9

**Comment by Bogdan107** [comment]
> Maybe you could override function `et_prerequisites` ... copy and paste ...

Proposed variant needs copying ~ 220 lines into plugin file.
Choosing between adding 220 lines and having additional question before evil twin attacks - I want to leave as is.

I propose alternative way:
1) add _AIRGEDDON_SPOOF_MAC_ADDRESS_AUTOMATIC_ configuration option;
2) import `spoof_mac_automatically.sh` plugin into airgeddon;
3) supress asking of mac-address spoofing before evil twin attacks, if option _AIRGEDDON_SPOOF_MAC_ADDRESS_AUTOMATIC_ activated.

**Comment by xpz3** [comment]
For a plugin, airgeddon won't be doing any changes to its code. That's the purpose of plugin itself. So if you don't prefer to override the function, maybe you can remove `spoof_mac_automatically_prehook_et_prerequisites()` and let the user decide if spoofing is needed for evil twin attacks.

**Comment by Bogdan107** [comment]
> For a plugin, airgeddon won't be doing any changes to its code. That's the purpose of plugin itself. So if you don't prefer to override the function, maybe you can remove `spoof_mac_automatically_prehook_et_prerequisites()` and let the user decide if spoofing is needed for evil twin attacks.

Fixed in [51e866c](https://github.com/Bogdan107/airgeddon_plugins/commit/51e866cad0163c9fe20f6b6e451e34069685b46f)

**Comment by OscarAkaElvis** [comment]
Ok, one last suggestion... airgeddon, as normal, after spoofing mac address, once the attack is finished, on function `restore_et_interface` is deleting the splitted interface and in this way the spoof is removed. It seems using your plugin, after spoofing mac, it is not restored. Only it is done on evil twin attacks because airgeddon is doing it as I explained. Maybe you should create a function to call it after finishing each attack to restore original mac or something like that. As I said, it's just a suggestion.

**Comment by Bogdan107** [comment]
> Maybe you should create a function to call it after finishing each attack to restore original mac or something like that.

fixed in [69ff76c](https://github.com/Bogdan107/airgeddon_plugins/commit/69ff76c8a5d77f41ca5e7e5ecbcb5983d265b647) (v0.2)
